SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestClientNetOwnership/SpatialTestNetOwnership.cpp

```diff
@@ -0,0 +1,151 @@
+ *  - Client 1 sends an RPC from the NetOwnershipCube at each test location.
+ * - Phase 1 test:
+ *  - All workers test that all sent RPCs have been received by the NetOwnershipCube.
+ *
+ * - Phase 2 setup
+ *  - The authoritative server sets the NetOwnershipCube's Owner to null.
+ *  - Client 1 sends an RPC from the NetOwnershipCube.
+ * - Phase 2 test:
+ *  - All workers test that the RPC sent was ignored.
+ * 
+ * - Test Cleanup
+ *  - The NetOwnershipCube is destroyed.
+ */
+
+ASpatialTestNetOwnership::ASpatialTestNetOwnership()
+	: Super()
+{
+	Author = "Andrei";
+	Description = TEXT("Test Net Ownership");
+}
+
+void ASpatialTestNetOwnership::BeginPlay()
+{
+	Super::BeginPlay();
+
+	// Step definition for Client 1 to send a Server RPC
+	FSpatialFunctionalTestStepDefinition ClientSendRPCStepDefinition;
+	ClientSendRPCStepDefinition.bIsNativeDefinition = true;
+	ClientSendRPCStepDefinition.TimeLimit = 5.0f;
+	ClientSendRPCStepDefinition.NativeStartEvent.BindLambda([this](ASpatialFunctionalTest * NetTest)
+		{
+			NetOwnershipCube->ServerIncreaseRPCCount();
+
+			FinishStep();
+		});
+
+	// Test Phase 1
+
+	// Server 1 spawns the NetOwnershipCube and registers it for auto-destroy.
+	AddStep(TEXT("SpatialTestNetOwnershipServerSpawnCube"), FWorkerDefinition::Server(1), nullptr, [this](ASpatialFunctionalTest* NetTest)
+		{
+			ANetOwnershipCube* NetOwnershipCube = GetWorld()->SpawnActor<ANetOwnershipCube>(FVector::ZeroVector, FRotator::ZeroRotator, FActorSpawnParameters());
+			RegisterAutoDestroyActor(NetOwnershipCube);
+
+			FinishStep();
+		});
+
+	AddStep(TEXT("SpatialTestNetOwnershipAllWorkersSetReference"), FWorkerDefinition::AllWorkers, nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
+		{
+			TArray<AActor*> NetOwnershipCubes;
+			UGameplayStatics::GetAllActorsOfClass(GetWorld(), ANetOwnershipCube::StaticClass(), NetOwnershipCubes);
+
+			// Make sure the cube is visible before trying to set the Test's variable.
+			if (NetOwnershipCubes.Num() != 1)
+			{
+				return;
+			}
+
+			NetOwnershipCube = Cast<ANetOwnershipCube>(NetOwnershipCubes[0]);
+
+			FinishStep();
+		}, 5.0f);
+
+	//  The authoritative server sets the owner of the NetOwnershipCube to Client's 1 PlayerController
+	AddStep(TEXT("SpatialTestNetOwnershipServerSetOwner"), FWorkerDefinition::AllServers, nullptr, [this](ASpatialFunctionalTest* NetTest)
+		{
+			if (NetOwnershipCube->HasAuthority())
+			{
+				APlayerController* PlayerController = Cast<APlayerController>(GetFlowController(ESpatialFunctionalTestWorkerType::Client, 1)->GetOwner());
+				NetOwnershipCube->SetOwner(PlayerController);
+			}
+
+			FinishStep();
+		});
+
+	// The locations where the NetOwnershipCube will be when Client 1 will send an RPC. These are specifically set to make the NetOwnershipCube's authoritative server change according to the BP_QuadrantZoningSettings.
```

Comment:
This is not future proof, as per the changes I will soon push to the GDK.
These changes will make the Cube join its owner's auth server.
In order to have the authority change, you have to move the player controller as well (which can be done by moving its possessed pawn).

Reply:
Thank you for letting me know, will change it now!